### ISSUE TITLE
[boost] cross-compile for QNX: gen "cmake_find_package" can't find libs

### ISSUE THREAD

**Issue by SSE4** [conan-community/community#78]
_From @Johnnyxy on August 28, 2018 9:11_

Hi there,
I am using Conan 1.6.1 on Windows (7; x64) with Boost 1.67.0 with the fancy pretty good "cmake_find_package"-generator. The best thing to seperate Conan and CMake and is much more traceable if something odd happens. The generator generates the Find-Module for Boost. It places the lib names provided by the recipe into the boost_LIBRARY_LIST variable.

`SET(boost_LIBRARY_LIST boost_wave boost_container boost_contract ...`

Now after that the libraries are searched by the CMake `find_library` command with the given library name from the list above.

Windows
---------
On Windows the files it has to find are named like this: boost_wave.lib boost_container.lib ...
`find_library` finds them and everything is good to go.

QNX
-----
On QNX the files it has to find are named like this: boost_wave.a boost_container.a ...
`find_library` cannot find them.

reason
-------
The reason behind this is propably that CMake's find_library does not search for `<name>.{a;so}` for the [QNX system](https://cmake.org/cmake/help/latest/variable/CMAKE_SYSTEM_NAME.html).
In the end find_library says that it cannot find e.g. _"boost_wave"_ because it did not search for _"boost_wave.a"_ just for _"libboost_wave.a"_ and maybe (I am unsure) just the plain name _"boost_wave"_.

see the find_library search modifying variables:
[CMAKE_FIND_LIBRARY_PREFIXES](https://cmake.org/cmake/help/latest/variable/CMAKE_FIND_LIBRARY_PREFIXES.html): lib
[CMAKE_FIND_LIBRARY_SUFFIXES](https://cmake.org/cmake/help/latest/variable/CMAKE_FIND_LIBRARY_SUFFIXES.html): .so;.a

workaround
------------
I added the following line directly in front of the _find_library_ command in the generated Findboost.cmake
`list (APPEND CMAKE_FIND_LIBRARY_PREFIXES "")`
so that the content of CMAKE_FIND_LIBRARY_PREFIXES is "lib;" (prefix "lib" and "") and find_library found everything.
But as the Find-file is generated this is no persistent solution.

So I added the same line to my CMakeLists.txt directly in front of
`find_package (boost 1.67.0 REQUIRED ...)`
and it worked.

aftermath
----------
As I am unsure who is _"responsible"_ for a clean solution. I opened a issue on gitlab at CMake. One CMake official replied that he never saw a system that does not use `lib<name>.{a;so}`. As I concur that this is odd for *nix systems but still a scenario.

One can find the CMake issue [here](https://gitlab.kitware.com/cmake/cmake/issues/18285#note_448792).

From my point of view I see nothing that stays against "no-prefix" as additional default entry in CMAKE_FIND_LIBRARY_PREFIXES.

_Copied from original issue: conan-community/conan-boost#148_

**Comment by SSE4** [comment]
hi @Johnnyxy 
unfortunately I have no access to QNX system to debug an issue.
do you know some easy way to get such an environment available, like some free Cloud VMs?

**Comment by Johnnyxy** [comment]
Hi @SSE4 
I don't think this has to do anything with QNX itself. As the problem is on CMake's side. CMake just assumes filename prefixes, postfixes and file extensions for certain systems (Windows: \<name>.{lib;dll}; Linux: lib\<name>.{a;so}).

I opened an [issue ](https://gitlab.kitware.com/cmake/cmake/issues/18285#note_448957) at CMake's bugtracker and got the following response from
Brad King: "I've never seen a platform that uses \<name>.{a,so} instead of lib\<name>.{a,so}.  I suspect that Boost's build system is incorrectly choosing library names when cross-compiling to QNX from Windows."

So I suspect you can test this with any library name on any ordinary Linux/Unix system. I think for Linux CMake does expect "lib\<name>.{a;so}" too. You could check the "CMAKE_FIND_LIBRARY_PREFIXES" CMake variable for your favourite Linux. It should just contain "lib".

To reproduce this rename a library of your choice to "myfancyname.a" and try to find it on your Linux with:
```cmake
set (MY_SEARCH_PATHS "PATH_TO_MY_LIBRARY" "...")
set (MY_LIB_NAME "myfancyname")
find_library (FOUND_LIBRARY NAME "${MY_LIB_NAME}" PATHS ${MY_SEARCH_PATHS} NO_DEFAULT_PATH NO_CMAKE_FIND_ROOT_PATH)
if (FOUND_LIBRARY)
    message (FATAL_ERROR "requested library found ({name: ${MY_LIB_NAME}}, {fullfilename: ${FOUND_LIBRARY}})")
else()
    message (FATAL_ERROR "requested library not found ({name: ${MY_LIB_NAME}})")
endif()
```

My two cents: The name scheme for Linux systems "lib\<name>.{...}" is just a convention not a technical necessity. So at least CMake should consider (with a lower priority of course) that there could be a library name without any prefixes.

Best regards :)

**Comment by SSE4** [comment]
@Johnnyxy this might be an issue we're not passing correct `target-os` value into `b2` in case of QNX.
it would be helpful if you can paste your conan profile here (output of `conan profile show default`).
it seems like default conan's `settings.yml` file has no QNX entry, and boost recipe also doesn't handle QNX at all.

**Comment by Johnnyxy** [comment]
@SSE4 as you had serval questions and I do not like unspecific ansers I will elaborate in detail my business with QNX.

CMAKE - find_library
----------------------
It is solely a problem with the find_library-mechanism CMake uses. The boost recipe renames the filenames of the libs from boost's `lib...gcc...x86...mt...whatever.a` to the `boost_modulename.a`. This is quite convenient as now we just could request the module name in CMake without thinking about what os, arch, etc.. Conan provides the correct libs and CMake should just use them.

The looked for prefixes just has to include an empty element. I added the following line for testing into the Conan generated (by "cmake_find_package"-generator) `Findboost.cmake` before the `find_library` command and CMake found the boost `<name>.a` successfully.
```cmake
set (CMAKE_FIND_LIBRARY_PREFIXES "lib" "")
```

There is an option in the conan recipe to rename the libraries after the build -> `skip_lib_rename=False`.
I one does not use this the result will be far more inconvenient than just adding the prefix-element to `CMAKE_FIND_LIBRARY_PREFIXES` by the generator.

Conan - default profile
-----------------------
I do not know what the profile has to do with that, but as you requested I will share it :).
Side note: Compiling for QNX Neutrino is done by cross-compiling on Windows/Linux for QNX Neutrino with a SDK provided by QNX. I compile only on Windows for QNX Neutrino at the moment.

Configuration for profile default:

[settings]
os=Windows
os_build=Windows
arch=x86_64
arch_build=x86_64
compiler=Visual Studio
compiler.version=14
build_type=Release
[options]
[build_requires]
[env]

QNX Neutrino
---------------
Just for clarification of my used nomenclature. The operating system is called `Neutrino`. So when I write QNX Neutrino then I mean the operating system. Normally it is abbreviated by "QNX" (Wiki: [eng](https://en.wikipedia.org/wiki/QNX); [rus](https://ru.wikipedia.org/wiki/QNX)). But that is only the company making the operating system.
If one wants to build for QNX operating system one has to by a license. With that license one can get access to specific SDKs. The SDK provides a QNX-modified GCC compiler called "qcc".
QNX SDK 6.5.0 -> C compiler "qcc"; C++ compiler "QCC"; version 4.4.2
QNX SDK 7.0.0 -> C compiler "qcc"; C++ compiler "q++"; version 5.4.0



Additionally the SDK provides the C/C++ runtime and some commonly used libraries for the targeted QNX Neutrino.
QNX SDK 6.5.0 -> 
- _gpp — GNU C++ lib (available only for x86)
- _cpp — Dinkum C++ lib (default)
- _cpp-ne — Dinkum C++ lib (no exceptions)
- _acpp — Dinkum Abridged C++ lib
- _acpp-ne — Dinkum Abridged C++ lib (no exceptions)
- _ecpp — Embedded Dinkum C++ lib
- _ecpp-ne — Embedded Dinkum C++ lib (no exceptions)

QNX SDK 7.0.0 ->
- _cxx — LLVM C++ library (default)
- _gpp — GNU C++ library



Boost - QNX
--------------
I built Boost for QNX 6.5.0 ([documentation](http://www.qnx.com/developers/docs/6.5.0/index.jsp?topic=%2Fcom.qnx.doc.neutrino_utilities%2Fq%2Fqcc.html)) with the following b2-line:
`
b2 toolset=qcc target-os=qnxnto threadapi=pthread link=static --with-system --with-thread --with-date_time --with-atomic --with-log define=BOOST_THREAD_DONT_USE_CHRONO cxxflags="-Vgcc_ntox86_gpp"
`
The --with- and define parameters where specific for my use case. I do not think they are necessary in general.
The significant parameter for the target to build the libs for is the `-V` parameter. [QNX 6.5.0](http://www.qnx.com/developers/docs/6.5.0/index.jsp?topic=%2Fcom.qnx.doc.neutrino_utilities%2Fq%2Fqcc.html) and [QNX 7.0.0](http://www.qnx.com/developers/docs/7.0.0/#com.qnx.doc.neutrino.utilities/topic/q/qcc.html) differs here.

[QNX 6.5.0 supports](http://www.qnx.com/developers/docs/6.5.0/index.jsp?topic=%2Fcom.qnx.doc.neutrino_utilities%2Fq%2Fqcc.html) the following targets out-of-the-box (you can omit the compiler version "4.4.4," part; see doc):
> 4.4.2,gcc_ntoarmle
4.4.2,gcc_ntoarmle_acpp-ne
4.4.2,gcc_ntoarmle_acpp
4.4.2,gcc_ntoarmle_cpp-ne
4.4.2,gcc_ntoarmle_cpp
4.4.2,gcc_ntoarmle_ecpp-ne
4.4.2,gcc_ntoarmle_ecpp
4.4.2,gcc_ntoarmle_gpp
4.4.2,gcc_ntoarmv7le
4.4.2,gcc_ntoarmv7le_acpp-ne
4.4.2,gcc_ntoarmv7le_acpp
4.4.2,gcc_ntoarmv7le_cpp-ne
4.4.2,gcc_ntoarmv7le_cpp
4.4.2,gcc_ntoarmv7le_ecpp-ne
4.4.2,gcc_ntoarmv7le_ecpp
4.4.2,gcc_ntoarmv7le_gpp
4.4.2,gcc_ntomipsbe
4.4.2,gcc_ntomipsbe_acpp-ne
4.4.2,gcc_ntomipsbe_acpp
4.4.2,gcc_ntomipsbe_cpp-ne
4.4.2,gcc_ntomipsbe_cpp
4.4.2,gcc_ntomipsbe_ecpp-ne
4.4.2,gcc_ntomipsbe_ecpp
4.4.2,gcc_ntomipsbe_gpp
4.4.2,gcc_ntomipsle
4.4.2,gcc_ntomipsle_acpp-ne
4.4.2,gcc_ntomipsle_acpp
4.4.2,gcc_ntomipsle_cpp-ne
4.4.2,gcc_ntomipsle_cpp
4.4.2,gcc_ntomipsle_ecpp-ne
4.4.2,gcc_ntomipsle_ecpp
4.4.2,gcc_ntomipsle_gpp
4.4.2,gcc_ntoppcbe
4.4.2,gcc_ntoppcbespe
4.4.2,gcc_ntoppcbespe_acpp-ne
4.4.2,gcc_ntoppcbespe_acpp
4.4.2,gcc_ntoppcbespe_cpp-ne
4.4.2,gcc_ntoppcbespe_cpp
4.4.2,gcc_ntoppcbespe_ecpp-ne
4.4.2,gcc_ntoppcbespe_ecpp
4.4.2,gcc_ntoppcbespe_gpp
4.4.2,gcc_ntoppcbe_acpp-ne
4.4.2,gcc_ntoppcbe_acpp
4.4.2,gcc_ntoppcbe_cpp-ne
4.4.2,gcc_ntoppcbe_cpp
4.4.2,gcc_ntoppcbe_ecpp-ne
4.4.2,gcc_ntoppcbe_ecpp
4.4.2,gcc_ntoppcbe_gpp
4.4.2,gcc_ntoshle
4.4.2,gcc_ntoshle_acpp-ne
4.4.2,gcc_ntoshle_acpp
4.4.2,gcc_ntoshle_cpp-ne
4.4.2,gcc_ntoshle_cpp
4.4.2,gcc_ntoshle_ecpp-ne
4.4.2,gcc_ntoshle_ecpp
4.4.2,gcc_ntoshle_gpp
4.4.2,gcc_ntox86        (default)
4.4.2,gcc_ntox86_acpp-ne
4.4.2,gcc_ntox86_acpp
4.4.2,gcc_ntox86_cpp-ne
4.4.2,gcc_ntox86_cpp
4.4.2,gcc_ntox86_ecpp-ne
4.4.2,gcc_ntox86_ecpp
4.4.2,gcc_ntox86_gpp

[QNX 7.0.0 supports](http://www.qnx.com/developers/docs/7.0.0/#com.qnx.doc.neutrino.utilities/topic/q/qcc.html) the following targets out-of-the-box (you can omit the compiler version "5.4.0," part; see doc):
> 5.4.0,gcc_ntoaarch64le
5.4.0,gcc_ntoaarch64le_cxx
5.4.0,gcc_ntoaarch64le_gpp
5.4.0,gcc_ntoarmv7le
5.4.0,gcc_ntoarmv7le_cxx
5.4.0,gcc_ntoarmv7le_gpp
5.4.0,gcc_ntox86        (default)
5.4.0,gcc_ntox86_64
5.4.0,gcc_ntox86_64_cxx
5.4.0,gcc_ntox86_64_gpp
5.4.0,gcc_ntox86_cxx
5.4.0,gcc_ntox86_gpp



I had to do some modifications but they are far from optimal. BZip2 and zlib dependencies are a problem.
1. QNX provides old zlib libraries/binaries (I think QNX_6.5.0 -> zlib_1.2.3; QNX_7.0.0 -> zlib_1.2.8).
2. QNX provides old bzip2 libraries/binaries.
My use case just needs boost header only which does not rely on linking with zlib/bzip2 so this is not a problem for me. I did not include any handling to build zlib/bzip2 for QNX in the `conanfile.py`.


Conan - Boost - QNX
----------------------
Conan does not have QNX support at all at the moment.
I just modified the `settings.yml`:
```
os:
    ...
    Neutrino: # QNX os is named "Neutrino"
        version: ["6.5.0", "7.0.0"]
compiler:
    ...
    qcc:
        version: ["4.4.2", "5.4.0"] # QNX 6.50 uses 4.4.2; 7.0.0 uses 5.4.0
        libcxx: [libstdc++, libc++] # QNX supports serval C/C++ runtime libs
```
and `conanfile.py` for version 1.67.0 from the [conan-community](https://github.com/conan-community/conan-boost):

modified the function `create_user_config_jam(self):`
```
    if (self.zip_bzip2_requires_needed):
            if (self.settings.os == u"Neutrino"):
                raise Exception(u"building or linking against zlib and bzip2 is not supported for QNX Neutrino at the moment")
            self.requires("bzip2/1.0.6@rms/adapterlib_stable")
```

added to the end of function `get_build_cross_flags(self):`
```
        # b2 toolset=qcc target-os=qnxnto threadapi=pthread link=static --with-system --with-thread --with-date_time --with-atomic --with-log define=BOOST_THREAD_DONT_USE_CHRONO cxxflags="-Vgcc_ntox86_gpp"
        if (self.settings.os == "Neutrino"):
            target = "qnxnto"
            compiler_target_runtime = ""
            if (not self.settings.compiler.runtime is None):
                if (self.settings.compiler.runtime == u"libc++"):
                    compiler_target_runtime = "_cxx"
                elif (self.settings.compiler.runtime == u"libstdc++"):
                    compiler_target_runtime = "_gpp"
            compiler_target = "gcc_nto%s%s" % (self.settings.arch, compiler_target_runtime)
            flags.append("cxxflags=\"-V%s\"" % compiler_target)

        if not target:
            raise Exception("Unknown target for %s" % self.settings.os)

        flags.append("target-os=%s" % target)
```

modified the function `create_user_config_jam(self):`
```
    if self.zip_bzip2_requires_needed:
            if (self.settings.os == u"Neutrino"):                
                contents += u"\nusing Neutrino provided bzip2 and zlib libraries"
            else:
                contents = "\nusing zlib : 1.2.11 : <include>%s <search>%s <name>%s ;" % (
```

modified the function `get_boostrap_toolset(self):`
```
        with_toolset = {"apple-clang": "darwin", "qcc": ""}.get(str(self.settings.compiler),
                                                                str(self.settings.compiler))
```

Comment
----------
As I have access to the compiler of QNX 6.5.0 (qcc/gcc 4.4.2) and 7.0.0 (qcc/gcc 5.4.0) I could help with compiling Boost and testing the `conanfile.py`. For QNX 6.6.0 I would have to check if it is available to me.
But be aware that testing the compiled code is almost not possible. I have no virtual machines at hand. I think one can buy VMs from QNX but I do not even know if there are VMs available for all QNX Neutrino supported architectures.



Best regards

**Comment by SSE4** [comment]
@Johnnyxy thanks, this is a really great introduction to the QNX itself and on how to compile a software for it as well
I have registered on QNX web-site and downloaded their cross-compiling tools.
I think I am going to address this issue in two steps:
1. add an official support for QNX Neutrino, QCC compiler and its standard libraries to the conan codebase
2. add QNX Neutrino support to the boost recipe
I'll keep you informed, probably will need your input on reviews and testing. thanks a lot!

**Comment by zmostafa** [comment]
@Johnnyxy I am trying to build boost also for QNX 7 ,do you use the same changes in the ercipes ? I am building boost 1.69.0 and took the changes for **get_build_cross_flags**() , and **get_boostrap_toolset**() ,and **I build bzip2 from source using qnx toolchain**.

I noticed that bootstrap.sh started with tool-set=gcc instead of qcc, and when I change the resipe file and make bootstrap starts with qcc, I get error /bootstrap/jam0 is not found.

I had to change the profile i used compiling bzip2 where I had defined all of my tools that matchs my target like following:
`
target_host=<home>/qnx700/host/linux/x86_64/usr/bin/aarch64-unknown-nto-qnx7.0.0
cc_compiler=gcc
cxx_compiler=g++
[env]
CHOST=$target_host
AR=$target_host-ar
AS=$target_host-as
RANLIB=$target_host-ranlib
LD=$target_host-ld
STRIP=$target_host-strip
CC=$target_host-$cc_compiler
CXX=$target_host-$cxx_compiler
`
I could not use this profile for compiling boost as boost uses parameters that are not valid for aarch64-unknown-nto-qnx7.0.0-g++, so i changed to qcc and q++,but at the end I noticed that this bin "i586-pc-nto-qnx7.0.0-ld" is used for linking,and got me confused.
How can you configure linking and assembler ... etc for qcc and q++ to have a similar profile like my first one ?

**Comment by Johnnyxy** [comment]
I have to use Boost too, but it is sufficient to use the ```header_only``` option from the boost recipe in my case. So the Boost recipe does not build bzip2 and zlib.
But I tested building Boost 1.69.0 (current version) along with bzip2 (1.0.6) and zlib2 (1.2.11) as dependencies and it worked.

Attachments
-------------
[qnx_7_boost_bzip2_zlib_and_res.zip](https://github.com/conan-community/community/files/2945076/qnx_7_boost_bzip2_zlib_and_res.zip)

I am using some modified bzip2/zlib/boost recipes for building. They are heavily based on conan-center/bintray recipes. I need to modify them as I cannot download the sources from the internet.

Both bzip2 and zlib have their sources along with the recipes. Only for the Boost recipe you have to decide how you get the sources (conanfile.py: line 130). I myself download them from a local network FTP server.

For Conan I have included everything you need to build for QNX 7.0 (Conan profile and CMake toolchain). I have prepared Conan profiles for serval architectures I have to support. You will find them in the attachment.

The following command should point you in the right direction to successfully build Boost with BZip2 and Zlib on QNX 7.0 (example is for ```aarch64le``` architecture and ```cxx``` runtime).
Modify the paths and os_build/os_arch and you are good to go.

```bat
mkdir build
cd build

conan install .. -s os_build="Windows" -s arch_build="x86_64" -s compiler="qcc" -s compiler.version="5.4" -s compiler.libcxx="cxx" -s build_type="Release" --update --json install.json --build --profile="./../res/conan/aarch64le-qnx-neutrino_7.0_qcc_5.4_cxx"
```

**Comment by SSE4** [comment]
need to try with conan 1.16 once it's released

**Comment by Johnnyxy** [comment]
@zmostafa 
Just for your information if you encounter the following case:

I tried to build Boost 1.69.0 and 1.70.0 with QNX SDP 7.0 with runtime cxx (LLVM backend). I can tell you this will not work. As there is a configuration error in the LLVM backend of the QCC (5.4.0).
The LLVM backend will disable certain compiler features which Boost uses but are actually present and working.

See here: https://github.com/boostorg/boost/issues/208#issuecomment-498182528

**Comment by SSE4** [comment]
blocked by boost bug: https://github.com/boostorg/config/issues/288

**Comment by SSE4** [comment]
closing as outdated
